packages/discord-request/src/queue.ts

```diff
@@ -0,0 +1,238 @@
+				retries < this.manager.config.retries
+			) {
+				return await this.#process(route, resource, init, retries + 1);
+			}
+
+			throw error;
+		} finally {
+			this.#shutdownSignal?.removeEventListener("signal", clearAbort);
+			clearAbort();
+		}
+
+		// Parse Rate Limit information
+		let retryAfter = 0;
+
+		const global = Boolean(response.headers.get("X-RateLimit-Global"));
+		const key = response.headers.get("X-RateLimit-Bucket");
+
+		const retry = response.headers.get("Retry-After");
+		const reset = response.headers.get("X-RateLimit-Reset-After");
+		const limit = response.headers.get("X-RateLimit-Limit");
+		const remaining = response.headers.get("X-RateLimit-Remaining");
+
+		this.reset = reset
+			? 1000 * Number(reset) + Date.now() + OFFSET
+			: Date.now();
+		this.limit = limit ? Number(limit) : Infinity;
+		this.remaining = remaining ? Number(remaining) : 1;
+
+		if (retry != null) {
+			retryAfter = Number(retry) * 1000 + OFFSET;
+		}
+
+		// Handle global rate limits
+		if (global && retryAfter > 0) {
+			this.manager.globalRequestCounter = 0;
+			this.manager.globalReset = Date.now() + retryAfter;
+		}
+
+		// Update bucket hashes as needed
+		const identifier = getRouteKey(init.method as RequestMethod, route);
+
+		if (key != null && key !== this.id) {
+			log(`Received new bucket hash. ${this.id} -> ${key}`);
+
+			this.manager.buckets.set(identifier, {
+				key,
+				lastRequest: Date.now(),
+			});
+		} else if (key != null && this.manager.buckets.has(identifier)) {
+			this.manager.buckets.set(identifier, {
+				key,
+				lastRequest: Date.now(),
+			});
+		}
+
+		// Successful Requests
+		if (response.ok) {
+			return response.json();
+		}
+
+		// Rate Limited Requests
+		if (response.status === 429) {
+			const rateLimitData = {
+				retryAfter,
+				bucket: this.id,
+				url: resource,
+				limit: this.limit,
+				route: route.path,
+				identifier: route.identifier,
+				global: global,
+				method: init.method as RequestMethod,
+			};
+
+			this.manager.onRateLimit?.(rateLimitData);
+
+			log(`Encountered 429 rate limit. ${JSON.stringify(rateLimitData)}.`);
+
+			// If caused by a sublimit, wait it out here so other requests on the route can be handled
+			await sleep(retryAfter, this.#shutdownSignal);
+
+			// Don't bump retries for a non-server issue (the request is expected to succeed)
+			return this.#process(route, resource, init, retries);
+		}
+
+		// If given a server error, retry the request
+		if (response.status >= 500 && response.status < 600) {
+			if (retries < this.manager.config.retries) {
+				return this.#process(route, resource, init, retries + 1);
+			}
+
+			throw new RequestError(
+				resource,
+				init,
+				response,
+				`Discord Server Error encountered: ${response.statusText}`
+			);
+		}
+
+		// Handle non-ratelimited bad requests
+		if (response.status >= 400 && response.status < 500) {
+			// If we receive this status code, it means the token we had is no longer valid.
```

Comment:
```suggestion
			// If we receive a 401 status code, it means the token we had is no longer valid.
```

this means 4XX in my head here